openwisp_monitoring/monitoring/charts.py

```diff
@@ -145,32 +146,78 @@ def deep_merge_dicts(dict1, dict2):
+    assert 'description' in chart_config
+    assert 'order' in chart_config
+    assert 'query' in chart_config
+    if chart_config['query'] is not None:
+        assert isinstance(chart_config['query'], dict)
+        assert 'influxdb' in chart_config['query']
+    else:
+        assert 'unit' in chart_config
+    if 'colorscale' in chart_config:
+        assert 'max' in chart_config['colorscale']
+        assert 'min' in chart_config['colorscale']
+        assert 'label' in chart_config['colorscale']
+        assert 'scale' in chart_config['colorscale']
+
+
 def get_chart_configuration():
     charts = deep_merge_dicts(DEFAULT_CHARTS, app_settings.ADDITIONAL_CHARTS)
     # ensure configuration is not broken
     for key, options in charts.items():
-        assert 'type' in options
-        assert 'title' in options
-        assert 'description' in options
-        assert 'order' in options
-        assert 'query' in options
-        if options['query'] is not None:
-            assert isinstance(options['query'], dict)
-            assert 'influxdb' in options['query']
-        else:
-            assert 'unit' in options
-        if 'colorscale' in options:
-            assert 'max' in options['colorscale']
-            assert 'min' in options['colorscale']
-            assert 'label' in options['colorscale']
-            assert 'scale' in options['colorscale']
+        _validate_chart_configuration(options)
     return charts
 
 
+def register_chart_configuration(chart_name, chart_config):
+    """
+    Registers a new chart configuration.
+    """
+    if not isinstance(chart_name, str):
+        raise ImproperlyConfigured('Chart name should be type "str".')
+    if not isinstance(chart_config, dict):
+        raise ImproperlyConfigured('Chart configuration should be type "dict".')
+    if chart_name in DEFAULT_CHARTS:
+        raise ImproperlyConfigured(
+            f'{chart_name} is an already registered Chart Configuration.'
+        )
+
+    _validate_chart_configuration(chart_config)
+    DEFAULT_CHARTS.update({chart_name: chart_config})
+    _register_chart_configuration_choice(chart_name, chart_config)
+
+
+def unregister_chart_configuration(chart_name):
+    if not isinstance(chart_name, str):
+        raise ImproperlyConfigured('Chart configuration name should be type "str"')
+    if chart_name not in DEFAULT_CHARTS:
+        raise ImproperlyConfigured(f'No such Chart configuation "{chart_name}"')
+    DEFAULT_CHARTS.pop(chart_name)
+    _unregister_chart_configuration_choice(chart_name)
+
+
+def _register_chart_configuration_choice(chart_name, chart_config):
+    name = chart_config.get('verbose_name', chart_name)
+    CHART_CONFIGURATION_CHOICES.append((chart_name, name))
+
+
+def _unregister_chart_configuration_choice(chart_config):
+    for index, (key, name) in enumerate(CHART_CONFIGURATION_CHOICES):
+        if key == chart_config:
+            CHART_CONFIGURATION_CHOICES.pop(index)
+            return
```

Comment:
The function should end with something like `raise ImproperlyConfigured(f'No such Chart Configuration {chart_name}')`

Reply:
`_unregister_chart_configuration_choice` is an internal method being called by `unregister_chart_configuration`. `unregister_chart_configuration` already takes care of the case you are pointing to, so adding it over here is not required.

Reply:
Hmm yes you are right and probably it's not necessary, I was just having a look at the implementation in openwisp-notifications which has it

https://github.com/openwisp/openwisp-notifications/blob/67e6dae18de579be0c60c599101272ba3bea59b3/openwisp_notifications/types.py#L74-L79

```
def _unregister_notification_choice(notification_type):
    for index, (key, name) in enumerate(NOTIFICATION_CHOICES):
        if key == notification_type:
            NOTIFICATION_CHOICES.pop(index)
            return
    raise ImproperlyConfigured(f'No such Notification Choice {notification_type}')
```